### DISCUSSION TITLE
Question about nats pub/sub

### DISCUSSION THREAD

**Discussion by karanpratapsingh** [nats-io/nats.go#945] [General]
Hey, I had a question, why don’t I see `Received: 1` as output?

```go
package main

import (
	"fmt"

	"github.com/nats-io/nats.go"
)

func main() {
	wait := make(chan bool)

	nc, err := nats.Connect(nats.DefaultURL)
	topic := "topic"

	if err != nil {
		fmt.Println(err)
	}
	nc.Publish(topic, []byte("1"))

	nc.Subscribe(topic, func(m *nats.Msg) {
		fmt.Printf("Received: %s\n", string(m.Data))
	})

	nc.Publish(topic, []byte("2"))
	nc.Publish(topic, []byte("3"))

	<-wait
}
```

Output:
```
Received: 2
Received: 3
```

**Comment by derekcollison** [comment]
When you subscribe you queue up the interest to the system, but you do that after you send []byte("1"), so when the system processes that message there is no interest in "topic".

**Comment by karanpratapsingh** [comment]
Thanks @derekcollison , makes perfect sense. I think [Basic Usage](https://github.com/nats-io/nats.go#basic-usage) in readme might need to be updated as I was running the example from there and couldn't figure out what was I doing wrong. Then I ran the basic usage from [nats.js](https://github.com/nats-io/nats.js#publish-and-subscribe) in node and it worked as subscribe came before publish (if this is as expected then feel free to close this issue).

Also, I just migrate from Kafka to Nats for one of my projects, It's pretty cool. Thanks!

**Comment by Vanlius** [comment]
> When you subscribe you queue up the interest to the system, but you do that after you send []byte("1"), so when the system processes that message there is no interest in "topic".

What about 'limitedPolicy' and 'WorkQueuePolicy' if The Retention in Jetstream is not InterestPolicy?

**Comment by derekcollison** [comment]
These are separate issues in my opinion.

JetStream allows you to persist messages, so the "interest" is JetStream itself, so users can come along at a later date and consume a message. JetStream can also be used to do global flow control, and if you need a globally ordered set of messages.
